wallet: avoid mixing different `OutputTypes` during coin selection

josibake opened the pull request:
# Concept

Following https://github.com/bitcoin/bitcoin/pull/23789, Bitcoin Core wallet will now generate a change address that matches the payment address type. This improves privacy by not revealing which of the outputs is the change at the time of the transaction in scenarios where the input address types differ from the payment address type. However, information about the change can be leaked in a later transaction. This proposal attempts to address that concern.

## Leaking information in a later transaction

Consider the following scenario:

![mix input types(1)](https://user-images.githubusercontent.com/7444140/158597086-788339b0-c698-4b60-bd45-9ede4cd3a483.png)

1. Alice has a wallet with bech32 type UTXOs and pays Bob, who gives her a P2SH address
2. Alice's wallet generates a P2SH change output, preserving her privacy in `txid: a`
3. Alice then pays Carol, who gives her a bech32 address
4. Alice's wallet combines the P2SH UTXO with a bech32 UTXO and `txid: b` has two bech32 outputs

From a chain analysis perspective, it is reasonable to infer that the P2SH input in `txid: b` was the change from `txid: a`. To avoid leaking information in this scenario, Alice's wallet should avoid picking the P2SH output and instead fund the transaction with only bech32 Outputs. If the payment to Carol can be funded with just the P2SH output, it should be preferred over the bech32 outputs as this will convert the P2SH UTXO to bech32 UTXOs via the payment and change outputs of the new transaction.

**TLDR;** Avoid mixing output types, spend non-default `OutputTypes` when it is economical to do so.

# Approach

`AvailableCoins` now populates a struct, which makes it easier to access coins by `OutputType`. Coin selection tries to find a funding solution by each output type and chooses the most economical by waste metric. If a solution can't be found without mixing, coin selection runs over the entire wallet, allowing mixing, which is the same as the current behavior.

I've also added a functional test (`test/functional/wallet_avoid_mixing_output_types.py`) and unit test (`src/wallet/test/availablecoins_tests.cpp`.

josibake committed:
refactor: store by OutputType in CoinsResult

Store COutputs by OutputType in CoinsResult.

The struct stores vectors of `COutput`s by `OutputType`
for more convenient access

josibake committed:
test: functional test for new coin selection logic

Create a wallet with mixed OutputTypes and send a volley of payments,
ensuring that there are no mixed OutputTypes in the txs. Finally,
verify that OutputTypes are mixed only when necessary.

MarcoFalke commented on src/wallet/spend.cpp:
```diff
@@ -767,10 +783,32 @@ static bool CreateTransactionInternal(
+    // First, do coin selection on a filtered subset of coins, filtered by OutputType
+    // Attempt with older output types first (LEGACY > P2SH > BECH32 > BECH32M)
+    std::optional<SelectionResult> result = [&] {
+	for (const OutputType& type : OUTPUT_TYPES) {
```
This assumes that `OUTPUT_TYPES` is sorted by age? Maybe add a comment there to ensure this doesn't break?

Also, this is the wrong indentation? Maybe run clang-format?

maflcko reviewed (commented):
I think an argument against this change is that any privacy leaking tx could very well be a coin-join between two wallets that use different output types. Though, I don't think any such coin-join software exists today, so until then it probably makes sense to use inputs of a single type. https://github.com/bitcoin/bitcoin/pull/24362#pullrequestreview-909069657 mentions that legacy inputs could preferably be used during low-fee periods to reduce the fees paid by the user. Though, if #24362 is merged, then it would probably not be worth it to optimize for fees.

josibake commented:
> I think an argument against this change is that any privacy leaking tx could very well be a coin-join between two wallets that use different output types. Though, I don't think any such coin-join software exists today, so until then it probably makes sense to use inputs of a single type. 

I agree, until mixed inputs in transactions are more common, this change should improve user privacy.

> https://github.com/bitcoin/bitcoin/pull/24362#pullrequestreview-909069657 mentions that legacy inputs could preferably be used during low-fee periods to reduce the fees paid by the user. Though, if https://github.com/bitcoin/bitcoin/pull/24362 is merged, then it would probably not be worth it to optimize for fees.

I prefer this approach over #24362 in order to keep the privacy benefits of matching change output addresses for all address types. I also see this as a small first step, where we can optimize for fees in a later PR if it is deemed useful/worth it.

josibake commented:
force pushed  `f900010` to `fdf1fa5` to fix clang-formatting and add comments per @MarcoFalke 's suggestion

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -186,6 +186,23 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
+                output_type = OutputType::BECH32M;
+            } else if (type == TxoutType::WITNESS_V0_KEYHASH) {
+                output_type = OutputType::BECH32;
+            } else if (type == TxoutType::SCRIPTHASH) {
+                output_type = OutputType::P2SH_SEGWIT;
```
In b32b52c3e7360ecb30e1a46dfa06669bc4fcd4b3 "wallet: avoid mixing ouput types in coin selection"

This is not guaranteed to be true. We could have a P2SH multisig and that would be `LEGACY`.

josibake replied:
is there a cleaner way to go map a `COutput` to an `OutputType`? adding lots of if/else checks to go from a `TxoutType` to an `OutputType` seems clunky to me

MarcoFalke replied:
Is there any need to map to output type? Why not use `TxoutType` directly?

achow101 replied:
> Is there any need to map to output type? Why not use `TxoutType` directly?

Yes, `P2SH_SEGWIT` is distinct from other P2SH which would be legacy.

josibake replied:
> This is not guaranteed to be true. We could have a P2SH multisig and that would be `LEGACY`.

so if a `P2SH_SEWGIT` and a `P2SH` multisig were spent together, when the UTXO is later spent it would reveal that some inputs were legacy multisig and others p2sh-segwit? 

As an alternative, would it be better to map `WITNESS_V0_SCRIPTHASH` to `P2SH` and map `SCRIPTHASH` and `PUBKEYHASH` to `LEGACY`?

achow101 replied:
By the time we get determining the output type, we already know that the output either belongs to the wallet, or is watch only. Thus we can do additional solver invocations for scripthash with the redeemScript in order to figure out the true type.

For watch only outputs, we may not be able to do that because not all of the scripts will necessarily be available. However we could still use solving data provided in `coinControl` to try to determine that.

josibake replied:
thanks for pointing me in the right direction! I'll take a stab at implementing this

josibake replied:
Now also considering the reedem script when a UTXO is `P2SH`

josibake commented:
> I would suggest running coin selection for each of the OutputGroup sets in parallel and then choosing the input set per the waste metric. Due to how the waste metric works, I'd expect this to lead to the older larger output types getting preferentially spent at lower feerates (and in the current feerate environment, usually first). What are your thoughts on that?

This makes a lot of sense. By running coin selection on each OutputGroup set you are preserving privacy by avoiding mixing the inputs, but also still choosing the best spend w.r.t to fees. Running them in parallel (as opposed to sequentially) also helps speed things up (which you also mentioned).

I'll work on refactoring to implement this.

luke-jr commented on src/wallet/spend.cpp:
```diff
@@ -767,10 +783,32 @@ static bool CreateTransactionInternal(
+    // First, do coin selection on a filtered subset of coins, filtered by OutputType
+    // Attempt with older output types first (LEGACY > P2SH > BECH32 > BECH32M)
+    std::optional<SelectionResult> result = [&] {
+        // This takes advantange of the fact that OUTPUT_TYPES is already sorted by age (LEGACY, P2SH, BECH32, BECH32M)
```
This assumes there is a strict preference for newer UTXO types. Rather, it should try to prefer using UTXOs that are not the user's preferred address type(s).

josibake replied:
sorry, forgot to respond to this! yes, rather than rely on a forced ordering of old to new, I'm implementing @Xekyo 's suggestion of running coin selection on each output group and then picking the solution with the best waste metric. If no solution is found, then run coin selection over all coins (current behavior).

josibake replied:
I've updated it now to not force an ordering on OutputType.

josibake committed:
wallet: run coin selection by `OutputType`

Run coin selection on each OutputType separately, choosing the best
solution according to the waste metric.

This is to avoid mixing UTXOs that are of different OutputTypes,
which can hurt privacy.

If no single OutputType can fund the transaction, then coin selection
considers the entire wallet, potentially mixing (current behavior).

This is done inside AttemptSelection so that all OutputTypes are
considered at each back-off in coin selection.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -186,6 +186,23 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
+                output_type = OutputType::P2SH_SEGWIT;
+            } else if (type == TxoutType::PUBKEYHASH) {
+                output_type = OutputType::LEGACY;
+            }
```
done!

josibake commented on src/outputtype.h:
```diff
@@ -22,6 +22,7 @@ enum class OutputType {
     BECH32M,
 };
 
+// Keep output types sorted by age, with older types first
```
removed, now that it is not dependent on ordering

murchandamus reviewed (commented):
Thanks for working on this, I like how you've split up availableCoins and how you have structured your tests. I have some ideas on some additional tests you might want to add, and a few questions.

josibake commented on test/functional/wallet_abandonconflict.py:
```diff
@@ -36,9 +36,9 @@ def run_test(self):
+        txB = self.nodes[1].sendtoaddress(alice.getnewaddress(), Decimal("10"))
+        txC = self.nodes[1].sendtoaddress(alice.getnewaddress(), Decimal("10"))
         self.sync_mempools()
         self.generate(self.nodes[1], 1)
```
interesting, is it by design that the waste metric does not consider unconfirmed change? a few ideas to address this:

1) have the waste metric also consider unconfirmed change
2) consider things like unconfirmed inputs when comparing solutions

for example, if solution A is to spend all confirmed inputs with a waste metric of X and solution B spends unconfirmed change with a waste metric of Y, prefer solution A even if X > Y.

Xekyo replied:
No, Bitcoin Core ignoring the feerate of parent transactions when using unconfirmed inputs is a bug, but one that turned out more difficult to fix than anticipated. I have started a branch (https://github.com/Xekyo/bitcoin/tree/ancestor-aware-funding) which I intend to push forward in collaboration with @glozow. We were stuck by not being able to get information on unconfirmed parent transactions in the wallet yet, and this should now be resolved by a node:wallet interface that provides miner score which @glozow came up with.

josibake replied:
This is better handled now by running by OutputType at each stage of back-off

glozow reviewed (commented):
Concept ACK

> Since the UTXO pools divided by type are much smaller, the backoff steps allowing for unconfirmed change to be used may come into play much more often. Then, since the waste metric does not yet account for the unconfirmed inputs, we may spend unconfirmed UTXOs more often.

> is it by design that the waste metric does not consider unconfirmed change?

Our preference for different coin selection solutions is spread out over multiple places in the code, including the waste metric and the sequence of `CoinEligibilityFilter`s used in each `AttemptSelection()` call. The way we encode our preference for confirmed UTXOs is by attempting coin selection with confirmed UTXOs first, and *only* trying with unconfirmed UTXOs if that fails.

Approach-wise, did you consider adding output type to `CoinEligibilityFilter` instead, and trying selection for each output type within `AttemptSelection()`? This way you try to do selection for every output type for confirmed UTXOs only, before you try unconfirmed.

Another possibility is adding the logic for preferring confirmed over unconfirmed UTXOs to `SelectionResult` and editing the `operator<` to account for it.

josibake commented on test/functional/wallet_avoidreuse.py:
```diff
@@ -293,7 +293,7 @@ def test_getbalances_used(self):
 
         # send multiple transactions, reusing one address
         for _ in range(101):
-            self.nodes[0].sendtoaddress(new_addr, 1)
```
this is related to your earlier comment (https://github.com/bitcoin/bitcoin/pull/24584#discussion_r836601796) regarding unconfirmed UTXOs being spent more often. In this test, after the first tx is sent, the next 100 txs build a chain where the unconfirmed change is being spent, eventually hitting the mempool chain limit of 25 and causing an error. By explicitly setting `fee_rate=1`, `nodes[0]` starts spending its confirmed coinbase UTXOs instead.

This also demonstrates your earlier point that this change is causing unconfirmed outputs to be spent more often

josibake replied:
I re-worked this to run coin selection by `OutputType` at each back-off, which fixed the preference for spending unconfirmed UTXOs. as such, I was able to leave the tests as is and removed the test changes from the PR.

mehdialiakbar31 replied:
> I re-worked this to run coin selection by `OutputType` at each back-off, which fixed the preference for spending unconfirmed UTXOs. as such, I was able to leave the tests as is and removed the test changes from the PR.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -81,11 +81,34 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *walle
+
+std::vector<COutput> AvailableCoins::all() {
+    std::vector<COutput> all;
+    all.reserve(Bech32m.size() + Bech32.size() + P2SH.size() + Legacy.size());
```
good catch, it should include `Other.size()`

josibake commented on src/wallet/spend.cpp:
```diff
@@ -816,9 +816,40 @@ static bool CreateTransactionInternal(
+
+    // Run coin selection on each OutputType and compute the Waste Metric
+    std::vector<SelectionResult> results;
+    if (auto result1{SelectCoins(wallet, availableCoins.Legacy, /* nTargetValue */ selection_target, coin_control, coin_selection_params)}) {
```
you're right, they can all be named `result` and I think that's better than having individual names for each run (`result_legacy', 'result_p2sh', etc)

josibake commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,214 @@
+        assert_equal(change, input_addrs[0])
+
+        self.generate(B, 1)
+        self.log.debug(
+            "Make the same payment again, with a higher feerate. "
+            "Check that Alice's wallet prefers bech32 UTXOs."
+        )
```
I think we are on the same page, it's just my wording in the comments isn't correct. Initially, I wrote the comments for when I was strictly preferring older types to newer types. I'll update these comments to be more accurate.

josibake commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,214 @@
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
+
+    def test_case_one(self, A, B):
```
good point, both nodes are starting with a clean chain (`setup_clean_chain = True`), and then node A mines 100 blocks in `run_test`.

Much better would be to explain this at the beginning of the test so it's clear what the respective UTXO sets for each node are.

josibake replied:
fixed

josibake commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,214 @@
+This test verifies that mixing different output types is avoided unless
+absolutely necessary. It does so by covering the following cases:
+
+    CASE 1: Alice mixed UTXOs (bech32, P2SH). She makes multiple payments
+            to Bob, where bech32 UTXOs are preferred at higher fee rates
+            and P2SH are preffered at lower fee rates.
+    CASE 2: Alice has mixed UTXOs (P2SH, bech32, legacy) and makes multiple
+            payments to Bob's bech32 address, where legacy is preferred at
+            low fee rates, P2SH at medium fee rates, and bech32 at high rates.
+    CASE 3: Alice has mixed UTXOs (legacy, P2SH, bech32) and makes multiple
+            payments to Bob, where legacy and P2SH are mixed at lower fee rates
+            and finally all UTXOs are mixed for the final payment.
```
Great points. I'll refactor the test to:

* Document the initial setup for each node
* Start with a larger, mixed UTXO set
* Be less brittle w.r.t to coin selection

Looking at the test again, I think I am trying to test for some things that would be better to test for in a unit test. Ideally, this test should create large UTXO sets for each node, spend back and forth at different fee rates and to different address types. The assertions should check that all inputs are of the same OutputType and finally test a scenario where the wallet can fund from any single OutputType and is forced to mix OutputTypes to fund the transaction

josibake replied:
this should be addressed now. I re-worked the test to make it less dependent on specific UTXOs and transactions and added a more detailed explanation.

josibake commented:
based on feedback so far (thanks, everyone!), I think next steps are:

* Better break out `P2SH`, vs lumping all `P2SH` with `P2SH_SEGWIT`
* Account for unconfirmed txs being preferred when picking the "best" solution
* Refactor functional tests to be less tightly coupled to coin selection and better documented
* Potentially move some tests out of functional into unit tests

fanquake commented on src/wallet/spend.cpp:
```diff
@@ -773,11 +814,42 @@ static bool CreateTransactionInternal(
+        result1->ComputeAndSetWaste(coin_selection_params.m_cost_of_change);
+        results.push_back(*result1);
+    };
+    if (auto result2{SelectCoins(wallet, availableCoins.P2SH, /* nTargetValue */ selection_target, coin_control, coin_selection_params)}) {
```
```suggestion
    if (auto result2{SelectCoins(wallet, availableCoins.P2SH, /*nTargetValue=*/selection_target, coin_control, coin_selection_params)}) {
```

Anywhere you are adding /touching named args, please use the [style from the developer docs](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c-named-arguments).

josibake committed:
refactor: use CoinsResult struct in SelectCoins

Pass the whole CoinsResult struct to SelectCoins instead of only a
vector. This means we now have to remove preselected coins from each
OutputType vector and shuffle each vector individually.

Pass the whole CoinsResult struct to AttemptSelection. This involves
moving the logic in AttemptSelection to a newly named function,
ChooseSelectionResult. This will allow us to run ChooseSelectionResult
over each OutputType in a later commit. This ensures the backoffs work
properly.

Update unit and bench tests to use CoinResult.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -189,7 +212,26 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
 
-            vCoins.emplace_back(COutPoint(wtx.GetHash(), i), wtx.tx->vout.at(i), nDepth, input_bytes, spendable, solvable, safeTx, wtx.GetTxTime(), tx_from_me);
+            std::vector<std::vector<uint8_t>> dummy;
+            TxoutType type{Solver(wtx.tx->vout[i].scriptPubKey, dummy)};
```
fixed

josibake commented:
re-worked this based on feedback and overall much happier with the approach:

* When output is P2SH, run the solver on the `redeemScript` to determine the true `OutputType`. This puts P2SH outputs which are multisig into the legacy bucket @achow101 
* Instead of running `SelectCoins` multiple times over each `OutputType`, I moved the logic into `AttemptSelection`. This ensures that all of the existing back-offs are respected, which addressed the issue regarding unconfirmed change @Xekyo @glozow 
* Refactored the test to be better documented and less brittle w.r.t coin selection
* Removed changes to existing functional tests as they are no longer failing after fixing the unconfirmed change issue. I do believe there is room for improvement if a test relies on specific behavior from coin selection, but that is out of the scope of this PR

josibake commented:
force pushed a few minor code style edits. also noticed the branch is failing to build after rebasing on master. seems related to the fuzzer, will troubleshoot in the morning.

LarryRuane commented:
Concept ACK

The second commit (68d96078714c1ed1804c293a2f05c3fb33092efa) doesn't compile due to the `allow_mixed_output_types` optional argument to `AttemptSelection()`. Fixed in the third commit, but it's good if each commit compiles.

I would favor making `allow_mixed_output_types` non-optional.

josibake commented:
> Concept ACK
> 
> The second commit ([68d9607](https://github.com/bitcoin/bitcoin/commit/68d96078714c1ed1804c293a2f05c3fb33092efa)) doesn't compile due to the `allow_mixed_output_types` optional argument to `AttemptSelection()`. Fixed in the third commit, but it's good if each commit compiles.

ah, good catch! i must have broken it when trying to break up the commits more cleanly. 

> I would favor making `allow_mixed_output_types` non-optional.

can you explain a bit more what you mean by this?

LarryRuane commented:
> can you explain a bit more what you mean by this?

It's just that I prefer code to be more explicit rather than hidden. Optional arguments are good if an existing function has many callers, and it would be disruptive to change all of them. In this case, all the calls (except one in src/bench) are in the same file (spend.cpp), so it wouldn't be much to update the callers. But this is just a suggestion.

josibake commented:
> It's just that I prefer code to be more explicit rather than hidden

this is a great point. ill update when fixing the compile issue

DaunthE1 reviewed (commented):
I don't understand?

DaunthE1 commented:
How can anyone contribute,if it doesn't understand?How can i have confidence if i am not being educated?How will i gain if i feel i that what i may be doing are of no use?

DaunthE1 commented:
Where would this authoraztion come from?The universe?or from "only your picked,funded,equipped,educated and favored english speaking peers novel friends"?1

josibake commented:
fixed the compilation issue by my moving the `allow_mixed_output_types` bool to the correct commit and made the boolean non-optional (thanks @LarryRuane). having the bool non-optional helps the readability of the logic in coin selection.

also put the function comment changes for `AttemptSelection` in the correct commit.

`git range-diff master 31895fb 12bb64f`

joshuadbryant1985 commented:
Your really smart about this stuff  I wish I could meet you and you teach
me

On Sat, Apr 23, 2022, 2:20 PM josie ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In test/functional/wallet_avoid_mixing_output_types.py
> <https://github.com/bitcoin/bitcoin/pull/24584#discussion_r856956306>:
>
> > +    CASE 1: Alice mixed UTXOs (bech32, P2SH). She makes multiple payments
> +            to Bob, where bech32 UTXOs are preferred at higher fee rates
> +            and P2SH are preffered at lower fee rates.
> +    CASE 2: Alice has mixed UTXOs (P2SH, bech32, legacy) and makes multiple
> +            payments to Bob's bech32 address, where legacy is preferred at
> +            low fee rates, P2SH at medium fee rates, and bech32 at high rates.
> +    CASE 3: Alice has mixed UTXOs (legacy, P2SH, bech32) and makes multiple
> +            payments to Bob, where legacy and P2SH are mixed at lower fee rates
> +            and finally all UTXOs are mixed for the final payment.
>
> this should be addressed now. I re-worked the test to make it less
> dependent on specific UTXOs and transactions and added a more detailed
> explanation.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/24584#discussion_r856956306>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYEWKKEWZUQE7Q37DJOXU6DVGQ5M7ANCNFSM5Q33BSMA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

LarryRuane reviewed (commented):
Concept, tested, and mostly code review ACK 12bb64f9fcd11207e5996e41204ab2ea822379c1
Suggestions are minor. I would like to get a better understanding of the algorithm by studying the functional test, so I may have a few more comments. But this looks very good! Thanks for hosting review club!

LarryRuane commented:
The description has a typo (**Aproach**). Also, for future reference, my understanding is that the description (initial comment) should be fairly concise and mostly plaintext, because it ends up in the git commit history. It's probably better to make the detailed description a separate comment. It's probably too late here, because if you added a comment with details, it would end up at the bottom.

Xekyo commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,171 @@
+    has_p2sh = False
+    has_bech32 = False
+
+    for addr in inputs:
+        if is_legacy_address(node, addr):
+            has_legacy = True
+        if is_p2sh_address(node, addr):
+            has_p2sh = True
+        if is_bech32_address(node, addr):
+            has_bech32 = True
```
I would have expected to count all five types (incl. bech32m and others), even if you don't explicitly create them here. This test would currently not fail if I Alice additionally sent Bob a P2TR output.

josibake replied:
good point, ill update this

murchandamus reviewed (commented):
I like the new approach much better.

It seems to me that there should be some tests for the new AvailableCoins struct that ensures that all types get sorted into the right buckets. For example I would have hoped to see all types getting generated and seeing that all buckets were populated. This only happens indirectly right now, by testing whether there are mixed inputs when sweeping the wallet.

Also, I would prefer if all four types were tested instead of just two.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -525,34 +618,36 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
         // confirmations on outputs received from other wallets and only spend confirmed change.
-        if (auto r1{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 6, 0), vCoins, coin_selection_params)}) return r1;
-        if (auto r2{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), vCoins, coin_selection_params)}) return r2;
+        if (auto r1{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 6, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/false)}) return r1;
```
I agree but decided to leave them alone since it's not really in scope for this PR and would have involved touching extra lines just to remove the number

josibake commented on src/wallet/spend.cpp:
```diff
@@ -626,11 +657,13 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, AvailableCoins
                 }
             }
         }
+        // If our wallet does not allow us to spend unconfirmed change, attempt to fund the transaction with mixed OutputTypes
+        if (auto r9{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) return r9;
```
my reasoning for starting it at the `r6` back-off is because that is also the first back-off where we allow partial spending of `OutputGroups`. so the status quo in the wallet is "spend unconfirmed change before relaxing privacy-preserving constraints."

you could argue that mixed outputs are not as bad as partially spending an output group w.r.t privacy, but i would still advocate for it being considered along with unconfirmed change, meaning if spending unconfirmed change is more efficient in terms of the waste metric, do that instead of mixing.

Xekyo replied:
I'm not sure that's an apt comparison. Address reuse is generally discouraged and hopefully happens seldom for most wallets. In addition, we already run opportunistic avoid-partial-spend in parallel for every single coin selection attempt and prefer its result as long as it is not more expensive even when not explicitly required.

This PR however follows a change in the wallet behavior that will generally match the output type of the recipient on every spend. We must expect wallets to frequently have outputs of different types as a result. This PR does not introduce avoidance of mixing as parallel coin selection results but rather requires it for almost all selection attempts. Splitting the UTXO pool into two or more sets doesn't half the size of the possible input set candidates, but reduces the combination space to tiny fractions.

I expect this already to cause a reduction in changeless transactions, but worse, I expect the change as proposed to considerably increase the use of unconfirmed inputs and thusly reduce the resulting transactions' reliability. Given the existing feerate estimation issue with unconfirmed inputs, this would more frequently cause the wallet to undershoot the target feerate, especially if we allow up to 12 unconfirmed inputs before considering mixing.

I'd prefer that mixing be allowed with r2, I'd be okay with r3. I might be convinced to introduce mixing on r4 via simulation results showing that coin selection results and reliability do not get significantly worse. My stance on this would generally be improved by us having finished ancestor aware fee estimation and the waste metric having been updated to account for automatic bumping of low feerate parent transactions.

I'd also be curious what others think about this.

Xekyo replied:
Let me add an example to illustrate my point.

Five UTXOs produce 31 non-empty combinations:
![image](https://user-images.githubusercontent.com/4060799/167263471-ebac8d8b-7699-411d-9128-422b626fabdb.png)

Two sets of three and two UTXOs only yield 7+3 combinations:
![image](https://user-images.githubusercontent.com/4060799/167263547-4929efa7-c83c-49ed-8ccb-53842f7e9d5c.png)
![image](https://user-images.githubusercontent.com/4060799/167263566-2255c81e-84a5-45a6-9242-b3ccfd708cf7.png)

josibake replied:
thanks for the detailed explanation and example. very good points about splitting the UTXO set. 

I've changed the logic and added a new back-off that allows mixing _before_ considering unconfirmed change. as such, you will still get the privacy improvement when spending from confirmed UTXOs. 

what's nice about how this is implemented is that we can easily move the logic up or down by changing the boolean so if we are able to fix the fee estimation issue and convince ourselves via simulations that there are not undesirable effects, we can easily move the restriction down at a later date by changing a boolean.

josibake commented:
thanks for the review, @Xekyo !

> it seems to me that there should be some tests for the new AvailableCoins struct that ensures that all types get sorted into the right buckets

I am working on a unit test for `GetAvailableCoins` which covers this. I think a unit test is more appropriate than a functional test in this case. I was planning to add the unit test in a follow-up PR, but also happy to add it here, so long as this PR isn't getting too big

> Also, I would prefer if all four types were tested instead of just two.

I'll update the functional test, because agreed, no reason not to

josibake commented:
thanks for the review @LarryRuane and suggestions on code style!

> be fairly concise and mostly plaintext, because it ends up in the git commit history.

completely forgot about this, thanks for the reminder!

josibake commented:
force pushed to address feedback from @Xekyo and @LarryRuane 

`git range-diff master 12bb64f f73e545`

josibake committed:
test: add unit test for AvailableCoins

test that UTXOs are bucketed correctly after
running AvailableCoins

josibake commented:
updated with a unit test for `GetAvailableCoins` and changed the logic to allow mixing `OutputTypes` before spending unconfirmed change (thanks @Xekyo for the detailed explanation on this), along with a few other stylistic changes.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -444,8 +444,14 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
+std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, const AvailableCoins& availableCoins,
                                const CoinSelectionParams& coin_selection_params)
+{
+    std::optional<SelectionResult> result = ChooseSelectionResult(wallet, nTargetValue, eligibility_filter, availableCoins.all(), coin_selection_params);
```
In f42db05a0ea8a4acfaf8e0117824cfd0f1d0547b "refactor: use AvailableCoins struct in SelectCoins"

The commit message mentions a newly introuduced method `ChooseSelectionAlgorithm`, but the code introduces `ChooseSelectionResult`.

Xekyo commented on src/wallet/spend.cpp:
```diff
@@ -601,34 +630,38 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, AvailableCoins
-        if (auto r2{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params)}) return r2;
+        // Also avoid mixing different UTXO OutputTypes to preserve privacy
+        if (auto r1{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 6, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/false)}) return r1;
+        if (auto r2{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/false)}) return r2;
+
+        // If we can't fund the tx from confirmed UTXOs of a single OutputType, allow mixed output types
+        if (auto r3{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) return r3;
```
In df6aa1d597b4de1f4707719b3ad5518a94645ee1 "wallet: run coin selection by `OutputType`"

AFAICT, this duplicates the selection attempts with segregated types for `CoinEligibilityFilter(1, 1, 0).

Either way, the selection with segregated types is attempted first and mixing only happens as a last resort. So, directly setting `allow_mixed_output_types=true` should result in equivalent back-off steps without repeating the segregated output attempts before allowing mixing. 

Not adding this line also obsoletes the following commit to fix the duplicate appearance of `r3`,  321a13e177dc79581ec896568ab8e7633d95b17f "refactor: update variable names".
 
 ```suggestion
        // If we can't fund the tx from confirmed UTXOs of a single OutputType, allow mixed output types
        if (auto r2{AttemptSelection(wallet, value_to_select, CoinEligibilityFilter(1, 1, 0), coins, coin_selection_params, /*allow_mixed_output_types=*/true)}) return r2;
 ```

josibake replied:
ah, good catch! `AttemptSelection` will only try with mixed inputs if it is unable to find at least one solution from segregated inputs and if mixing is allowed. So yes, running `r2` with `allow_mixed_inputs=true` achieves want we want without needing the duplicated r3

Xekyo commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,195 @@
+    return False
+
+
+def is_p2sh_address(node, addr):
+    """Check if an address contains a P2SH output."""
+    addr_info = node.getaddressinfo(addr)
+    return (
+        addr_info['isscript'] and
+        not addr_info['iswitness']
+    )
+
+
+def is_legacy_address(node, addr):
+    """Check if an address contains a legacy output."""
+    addr_info = node.getaddressinfo(addr)
+    return (not addr_info['isscript'] and not addr_info['iswitness'])
```
In 784b03888e70a4744e7b387576f874ac18339dbb "test: functional test for new coin selection logic "

Not sure if this is relevant, but I think this may be inconsistent with the categorization pointed out by @achow101 previously, which IIUC prescribes that non-segwit P2SH addresses should be labeled as "legacy".

josibake replied:
it is inconsistent in the general sense, but for the test it is correct in that we are only generating `p2sh-segwit` outputs. i'll update the comment in the function, tho, to avoid any confusion

murchandamus reviewed (commented):
Looks pretty good, I think I'd consider this ready if the unnecessary additional back-off were removed.

josibake commented on src/wallet/spend.h:
```diff
@@ -35,15 +35,42 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* walle
  * to list all available coins (e.g. listunspent RPC) while not intending to fund a transaction.
  */
-void AvailableCoinsListUnspent(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+void AvailableCoinsListUnspent(const CWallet& wallet, AvailableCoins& coins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
```
originally, i tried to match whatever name was being used to minimize the diff, but looking at this again i think having a consistent naming of `available_coins` throughout is much better (despite a slightly larger diff)

josibake commented:
force-pushed for rebase (pretty big rebase following #25005) and also to address feedback from @Xekyo 

a few notes:
- after #25005 , it made more sense to leave the function named `AvailableCoins` (as opposed to changing it to `GetAvailableCoins`)
- `available_coins` is the preferred variable name throughout (vs `coin`, `available`, `availableCoins`, `vCoins`, etc)
- dropped a refactor commit after removing the extraneous step in the back-offs 
- fixed the functional test to be more clear and removed the unconfirmed tx test as mixing is allowed before spending unconfirmed change
- fixed an error in `bench/coin_selection.cpp` after making `allow_mixed_output_types` boolean required for the function call

furszy reviewed (commented):
Some initial, code-only, feedback so we can eliminate duplicated operations/code across the different vectors calls:

What about using a single `map<OutputType, std::vector>` instead of have one vector definition per output type and encapsulate the vector push, shuffle and erase inside the `CoinsResult::push_back`, `CoinsResult::shuffle` and `CoinsResult::erase`  methods?

Changes:

For d8cd07ec —> https://github.com/furszy/bitcoin-core/commit/d28e2cacae5a12b34c1f87de7815462e6380b779


For 38295279 —> https://github.com/furszy/bitcoin-core/commit/a5d4461eb30ad2da63a30a103ec16853689e3d38


For 27195339 —> https://github.com/furszy/bitcoin-core/commit/49a045ac91c48a94573bde937d99e78b13167f09


For eca95887 —> https://github.com/furszy/bitcoin-core/commit/f3d02885ec8a02dde5cf1721602a846bfbe0dfdb

—

Then another point is that the bench binary isn’t compiling:
 Need to add the new flag `allow_mixed_output_types` to `src/bench/coin_selection.cpp`, line 78.

josibake commented on src/wallet/spend.cpp:
```diff
@@ -204,9 +230,47 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+            if (wtx.tx->vout[i].scriptPubKey.IsPayToScriptHash()) {
+                CScript redeemScript;
+                CTxDestination destination;
+                ExtractDestination(wtx.tx->vout[i].scriptPubKey, destination);
```
im thinking just print a log and continue if we can't extract a destination?

josibake commented on src/wallet/spend.cpp:
```diff
@@ -204,9 +230,47 @@ CoinsResult AvailableCoins(const CWallet& wallet,
+                CTxDestination destination;
+                ExtractDestination(wtx.tx->vout[i].scriptPubKey, destination);
+                const CScriptID& hash = CScriptID(std::get<ScriptHash>(destination));
+                provider->GetCScript(hash, redeemScript);
```
added a check on `spendable` before looking up the CScript (since it doesn't make sense to try and find the script if it's not spendable by us) and handled the case where ExtractDest fails or GetScript fails with a `continue`.

josibake commented:
small fix-up on the functional test removing some unneeded lines

josibake commented:
rebased master and force pushed to fix a ci issue

LarryRuane commented on src/wallet/spend.h:
```diff
@@ -72,35 +96,52 @@ const CTxOut& FindNonChangeParentOutput(const CWallet& wallet, const COutPoint&
+ * returns                               If successful, a SelectionResult containing the input set
+ *                                       If failed, a nullopt
+ */
+std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, const CoinsResult& available_coins,
+                        const CoinSelectionParams& coin_selection_params, bool allow_mixed_output_types);
```
clang-format
```suggestion
std::optional<SelectionResult> AttemptSelection(
    const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, const CoinsResult& available_coins,
    const CoinSelectionParams& coin_selection_params, bool allow_mixed_output_types);
```

josibake replied:
i tried to avoid changing the formatting as much as possible to minimize the diff on this PR, so im inclined to leave it as is, but happy to change if others have an opinion about it.

not sure what the recommended best practice here is :man_shrugging:

LarryRuane commented on src/wallet/spend.h:
```diff
@@ -72,35 +96,52 @@ const CTxOut& FindNonChangeParentOutput(const CWallet& wallet, const COutPoint&
+ *                                       If failed, a nullopt
  */
-std::optional<SelectionResult> AttemptSelection(const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, std::vector<COutput> coins,
+std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, const std::vector<COutput>& available_coins,
                         const CoinSelectionParams& coin_selection_params);
```
clang-format
```suggestion
std::optional<SelectionResult> ChooseSelectionResult(
    const CWallet& wallet, const CAmount& nTargetValue, const CoinEligibilityFilter& eligibility_filter, const std::vector<COutput>& available_coins,
    const CoinSelectionParams& coin_selection_params);
```

LarryRuane commented on src/wallet/spend.h:
```diff
@@ -72,35 +96,52 @@ const CTxOut& FindNonChangeParentOutput(const CWallet& wallet, const COutPoint&
  *                                     If failed, a nullopt.
  */
-std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vector<COutput>& vAvailableCoins, const CAmount& nTargetValue, const CCoinControl& coin_control,
+std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& available_coins, const CAmount& nTargetValue, const CCoinControl& coin_control,
                  const CoinSelectionParams& coin_selection_params) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
```
clang-format
```suggestion
std::optional<SelectionResult> SelectCoins(
    const CWallet& wallet, CoinsResult& available_coins, const CAmount& nTargetValue, const CCoinControl& coin_control,
    const CoinSelectionParams& coin_selection_params) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
```

LarryRuane commented on src/wallet/test/availablecoins_tests.cpp:
```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
```suggestion
// file COPYING or https://www.opensource.org/licenses/mit-license.php.
```

josibake replied:
huh, interesting. seems like we have a mix of http and https throughout the codebase in the license header (mental note to fix this in a follow up)

LarryRuane commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,189 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
```suggestion
# file COPYING or https://www.opensource.org/licenses/mit-license.php.
```

LarryRuane commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,189 @@
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test output type mixing during coin selection
+
+A wallet may have different type UTXOs to choose from during coin selection,
```
```suggestion
A wallet may have different types of UTXOs to choose from during coin selection,
```

LarryRuane reviewed (commented):
Code review ACK 838685a4d018961dcd8315a59028aaa262b61916
Looks very good. My suggestions are all nits. I don't really understand the wallet code well; my review is pretty superficial so don't put much weight on my ACK.

1440000bytes commented:
> Leaking information in a later transaction

Concept ACK for fixing this

> Accumulating older OutputType UTXOs

Not sure about this because average users will use bech32/bech32m addresses. If some users prefer other addresses then it should be okay to accumulate other types in the wallet.

Still need to review and test the approach. My first reaction after looking at the diagram was if the transaction has mixed inputs, will it be okay to use mixed outputs?

![mixinmixout](https://user-images.githubusercontent.com/94559964/175802998-d7a3c840-5754-4575-8902-88c90fb8ab7d.png)

Or never used mixed inputs in a transaction?

achow101 commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,189 @@
+        self.generate(A, 1)
+
+        self.log.info("Sending payments from B to A")
+        for v in generate_payment_values(10, 20):
```
In ffcc600156e45968b47586e6758f107a56d6b80e "test: functional test for new coin selection logic"

How can we be sure that `v` is never greater than 10 (and thus result in mixing input types)?

achow101 commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,189 @@
+    return False
+
+
+def is_p2sh_segwit_address(node, addr):
+    """Check if an address contains a P2SH-Segwit outputself.
+       Note: this function does not actually determine the type
+       of P2SH output, but is sufficient for this test in that
+       we are only generating P2SH-Segwit outputs.
+    """
+    addr_info = node.getaddressinfo(addr)
+    return (
+        addr_info['isscript'] and
+        not addr_info['iswitness']
+    )
```
In ffcc600156e45968b47586e6758f107a56d6b80e "test: functional test for new coin selection logic"

The easiest way to check for address type is to check what the descriptor starts with. `sh(wpkh(` or `sh(wsh(` is p2sh-segwit, `wpkh(` or `wsh(` is bech32, `tr(` is bech32m, and `pkh(` is legacy.

josibake replied:
nice! this test only runs with `--descriptors` so ill update the `is_<addrtype>` functions

achow101 reviewed (commented):
It would be nice to have a test for the backoff behavior where we don't do mixing of 6 conf funds.

Xekyo commented on src/wallet/spend.h:
```diff
@@ -105,26 +105,42 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
+ * param@[in]  wallet                    The wallet which provides solving data for the coins
+ * param@[in]  nTargetValue              The target value
+ * param@[in]  eligilibity_filter        A filter containing rules for which coins are allowed to be included in this selection
+ * param@[in]  available_coins           The vector of coins available for selection prior to filtering
```
Consistency nit: In 3520cf9d76d24e43b3ba08fe03687ee1b2e2368e (refactor: use AvailableCoins struct in SelectCoins) 

As below, should perhaps be referred to as the struct?

```suggestion
 * param@[in]  available_coins           The struct of coins, organized by OutputType, available for selection prior to filtering
```

josibake commented on src/wallet/spend.cpp:
```diff
@@ -383,20 +453,55 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
+
+    // If we can't fund the transaction from any individual OutputType, run coin selection
+    // over all available coins, else pick the best solution from the results
+    std::optional<SelectionResult> result = [&] {
```
it was originally a lambda, but i do think it's more legible without it. also, removing the lambda also takes care of the shadowing

josibake commented on src/wallet/spend.cpp:
```diff
@@ -383,20 +453,55 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
+    std::optional<SelectionResult> result = [&] {
+        if (results.size() == 0) {
+            if (allow_mixed_output_types) {
+                if (auto result{ChooseSelectionResult(wallet, nTargetValue, eligibility_filter, available_coins.all(), coin_selection_params)}) {
```
fixed by removing the lambda

josibake commented on src/wallet/spend.cpp:
```diff
@@ -488,13 +592,13 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, const std::vec
-        else
-            ++it;
+    // remove preset inputs from coins so that Coin Selection doesn't pick them.
+    if (coin_control.HasSelected()) {
+        available_coins.legacy.erase(remove_if(available_coins.legacy.begin(), available_coins.legacy.end(), [&](const COutput& c) { return preset_coins.count(c.outpoint); }), available_coins.legacy.end());
+        available_coins.P2SH_segwit.erase(remove_if(available_coins.P2SH_segwit.begin(), available_coins.P2SH_segwit.end(), [&](const COutput& c) { return preset_coins.count(c.outpoint); }), available_coins.P2SH_segwit.end());
+        available_coins.bech32.erase(remove_if(available_coins.bech32.begin(), available_coins.bech32.end(), [&](const COutput& c) { return preset_coins.count(c.outpoint); }), available_coins.bech32.end());
+        available_coins.bech32m.erase(remove_if(available_coins.bech32m.begin(), available_coins.bech32m.end(), [&](const COutput& c) { return preset_coins.count(c.outpoint); }), available_coins.bech32m.end());
+        available_coins.other.erase(remove_if(available_coins.other.begin(), available_coins.other.end(), [&](const COutput& c) { return preset_coins.count(c.outpoint); }), available_coins.other.end());
     }
```
agree that this is not the most readable :sweat_smile: , however i really like @furszy 's suggestion for refactoring this: move the shuffle and erase functions into the `CoinsResult` struct. i didn't take his suggestion in this PR because it requires a bigger refactor than the scope of this PR, but i plan to address it in a follow-up.

for reference, it would look like this: https://github.com/furszy/bitcoin-core/commit/a5d4461eb30ad2da63a30a103ec16853689e3d38

josibake commented on src/wallet/spend.h:
```diff
@@ -34,14 +34,38 @@ struct TxSize {
+ * clear(), size() methods are implemented so that one can interact with
+ * the AvailableCoins struct as if it were a vector
+ */
 struct CoinsResult {
```
`CoinsResult` was introduced in https://github.com/bitcoin/bitcoin/pull/25005 , which changed `AvailableCoins` to return a struct instead of populating a vector and was introduced here after a rebase.

Personally, I prefer `AvailableCoins` as the name of the struct (and renaming the `AvailableCoins` function to `GetAvailableCoins`), but changing the name again in this PR felt a little silly and would have expanded the diff on this PR. I have a follow-up PR in mind which incorporates some feedback from @furszy regarding the `CoinsResult` struct, and I think that would be a better place to address the naming

josibake commented:
> Not sure about this because average users will use bech32/bech32m addresses. If some users prefer other addresses then it should be okay to accumulate other types in the wallet.

regardless of what the wallet default address type is, since https://github.com/bitcoin/bitcoin/pull/23789 , a user will accumulate non-default address types in their wallet based on the change address matching logic. this is the primary motivation for this PR. perhaps a better phrasing in the description would be "non-default" (instead of older types).

josibake commented:
force pushed for rebase; `git range-diff master 7f12300 2e30c09`

1440000bytes commented:
> > Not sure about this because average users will use bech32/bech32m addresses. If some users prefer other addresses then it should be okay to accumulate other types in the wallet.
> 
> regardless of what the wallet default address type is, since #23789 , a user will accumulate non-default address types in their wallet based on the change address matching logic. this is the primary motivation for this PR. perhaps a better phrasing in the description would be "non-default" (instead of older types).

It happens only when user sends bitcoin to non-default type. This was my point in the quoted part of comment.

josibake commented:
force-pushed for a rebase; `git range-diff master 2e30c09 6530d19`

Xekyo commented on src/wallet/test/availablecoins_tests.cpp:
```diff
@@ -0,0 +1,105 @@
+        CCoinControl dummy;
+        FeeCalculation fee_calc_out;
+        {
+            constexpr int RANDOM_CHANGE_POSITION = -1;
+            std::optional<CreatedTransactionResult> txr = CreateTransaction(*wallet, {recipient}, RANDOM_CHANGE_POSITION, error, dummy, fee_calc_out);
```
No change needed: Don't quite understand why the `RANDOM_CHANGE_POSITION` is defined specifically like this. Wouldn't this just work?

```suggestion
            std::optional<CreatedTransactionResult> txr = CreateTransaction(*wallet, {recipient}, /*change_pos=*/-1, error, dummy, fee_calc_out);
```

josibake replied:
iirc, i copied this from other tests which use `CreateTransaction`. personally, i prefer it to just `-1` because it's more clear to me having a `NAME_CONSTANT`, but i don't have a super strong feelings about it

murchandamus reviewed (approved):
ACK 6530d1950e6f6d5e1e64a2d59f244bb8c6b3996a

Got a few nits pertaining to comments and descriptions mostly. The code appears ready to ship. I think that some readability improvements could be made in a follow-up especially for instances where the same operations is performed for each output type. Basically the idea that @LarryRuane suggested would also apply where `AttemptSelection(…)` is called. The output types could be put in a list and the `AttemptSelection(…)` call could be made for each of the listed output types.

aureleoules commented:
I believe this PR conflicts with master silently?
I cannot compile the code when rebasing with master but do not get any merge conflicts.
It looks like the issue comes from the last commit 6530d1950e6f6d5e1e64a2d59f244bb8c6b3996a.

furszy commented on src/wallet/test/availablecoins_tests.cpp:
```diff
@@ -0,0 +1,105 @@
+    //   2. One UTXO from the change, due to payment address matching logic
+
+    // Bech32m
+    wallet->GetNewDestination(OutputType::BECH32M, "", dest, error);
```
Need to adapt this (and the ones below) to the new result class.

```
const auto& op_dest = wallet->GetNewDestination(OutputType::BECH32M, "");
BOOST_ASSERT(op_dest.HasRes());
const CTxDestination& dest = op_dest.GetObj();
```

aureleoules commented on src/wallet/spend.cpp:
```diff
@@ -379,20 +450,51 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
+                return result;
+            }
+        }
+        return std::optional<SelectionResult>();
```
nit
```suggestion
        return std::nullopt;
```

josibake replied:
I think I originally tried this and ran into strange behavior. Also, in other places an empty `SelectionResult` is returned as opposed to a nullopt. Going to leave this for now, but might be worth taking a second look at and perhaps using the newly created `BResult` class

achow101 replied:
It shouldn't be a problem to return `std::nullopt`, but the current code is equivalent anyways. The default constructor for a `std::optional` is to make a nullopt object.

aureleoules reviewed (commented):
ACK 71d1d13627ccd27319f347e2d8167c8fe8a433f4.

I believe the `CoinsResult` struct is much easier to work with than the previous coin vector.
This PR does seems to fix the privacy issue explained in the description.

kristapsk commented:
Concept ACK

josibake commented on src/wallet/spend.cpp:
```diff
@@ -379,20 +450,51 @@ std::vector<OutputGroup> GroupOutputs(const CWallet& wallet, const std::vector<C
+        }
+        return std::optional<SelectionResult>();
+    };
+    std::optional<SelectionResult> result{*std::min_element(results.begin(), results.end())};
+    return result;
```
will add if I need to touch this file again

josibake commented on test/functional/wallet_avoid_mixing_output_types.py:
```diff
@@ -0,0 +1,176 @@
+            ],
+        ]
+
+    def skip_test_if_missing_module(self):
+        self.skip_if_no_wallet()
```
I _think_ `skip_if_no_sqlite()` is sufficient for this? will add if i need to touch the functional test file again

achow101 commented:
re-ACK 71d1d13627ccd27319f347e2d8167c8fe8a433f4

1440000bytes reviewed (changes requested):
Approach NACK

I was not able to achieve the privacy mentioned in PR description. Steps to reproduce:

1. Create a wallet w1, send 0.001 four times from signet faucet to bech32 addresses, wait for confirmation and spend more than 0.002 to a legacy address.
2. There will be 2 UTXO in wallet: 1 legacy and other bech32
3. Send more than 0.001 to a bech32 address



Example:

https://mempool.space/signet/tx/6bb60f7dd4171d915b615030bc3cd2af785188834f38f391aa6b28e6a5359ecb:0

https://mempool.space/signet/tx/2cd2821390126c5407f29be8669c612a4092d0bc11224811a0752ff7fa31d43a

---

This is a valid argument and there can be multiple interpretations even without coinjoin: https://github.com/bitcoin/bitcoin/pull/24584#pullrequestreview-911639998

Also still curious about the two questions/solutions I had about solving this problem in https://github.com/bitcoin/bitcoin/pull/24584#issuecomment-1166432804

achow101 commented:
> Approach NACK
> 
> I was not able to achieve the privacy mentioned in PR description. Steps to reproduce:
> 
>     1. Create a wallet w1, send 0.001 four times from signet faucet to bech32 addresses, wait for confirmation and spend more than 0.002 to a legacy address.
> 
>     2. There will be 2 UTXO in wallet: 1 legacy and other bech32
> 
>     3. Send more than 0.001 to a bech32 address
> 
> 
> Example:
> 
> https://mempool.space/signet/tx/6bb60f7dd4171d915b615030bc3cd2af785188834f38f391aa6b28e6a5359ecb:0
> 
> https://mempool.space/signet/tx/2cd2821390126c5407f29be8669c612a4092d0bc11224811a0752ff7fa31d43a

I don't see how this example tests what is described in the OP? You ended up making a transaction that has mixed inputs, but the that's because those are the only funds in your wallet. The point of this PR is to *avoid* mixing input types, not *never* mix input types.

josibake commented:
hey @1440000bytes , thanks for testing! as mentioned, the goal is not to _never_ mix; the goal is to avoid mixing if possible. I've updated the OP to explicitly state this. also, take a look at the functional test, as this demonstrates the desired behavior: creating a wallet with mixed UTXO types, funding several transactions without mixing, and finally emptying the wallet (which is impossible to do without mixing).

> This is a valid argument and there can be multiple interpretations even without coinjoin: https://github.com/bitcoin/bitcoin/pull/24584#pullrequestreview-911639998

while there can be other interpretations, the most likely interpretation today is that the non-default input in the subsequent transaction is the change from the previous transaction (and in the majority of cases, this interpretation is correct).

> Also still curious about the two questions/solutions I had about solving this problem in https://github.com/bitcoin/bitcoin/pull/24584#issuecomment-1166432804

i responded here: https://github.com/bitcoin/bitcoin/pull/24584#issuecomment-1177991068 , please let me know which parts of your original question you feel are unaddressed.

1440000bytes commented:
How about insufficient funds error message or use mixed output types for change? This could happen even when some transactions are unconfirmed and confirmed UTXOs are of different types

Xekyo commented:
> How about insufficient funds error message or use mixed output types for change? This could happen even when some transactions are unconfirmed and confirmed UTXOs are of different types

An advanced user that cares so much about not mixing inputs is probably already using coin control to manually select inputs. I think this PR is a reasonable middle-path of rolling out a privacy improvement while minimizing negative impact on the UX. Adding cases in which transaction building fails while there are sufficient funds available as the default behavior would be an unreasonable UX disimprovement. I would consider the addition of options for more strict behavior to be out of scope for this PR.

Xekyo commented on src/wallet/test/availablecoins_tests.cpp:
```diff
@@ -0,0 +1,105 @@
+    BOOST_CHECK_EQUAL(available_coins.bech32.size(), 2U);
+
+    // P2SH-SEGWIT
+    dest = wallet->GetNewDestination(OutputType::P2SH_SEGWIT, "");
```
Nit:
```suggestion
    dest = wallet->GetNewDestination(OutputType::P2SH_SEGWIT, "");
    BOOST_ASSERT(dest.HasRes());
```

murchandamus reviewed (approved):
reACK 71d1d13627ccd27319f347e2d8167c8fe8a433f4 via `git range-diff master 6530d19 71d1d13`

LarryRuane commented on src/wallet/spend.cpp:
```diff
@@ -79,6 +79,32 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction &tx, const CWallet *walle
     return CalculateMaximumSignedTxSize(tx, wallet, txouts, coin_control);
 }
 
+uint64_t CoinsResult::size() const
```
commit 2e67291ca3ab2d8f498fa910738ca655fde11c5e refactor: store by OutputType in CoinsResult
nit
```suggestion
size_t CoinsResult::size() const
```

LarryRuane reviewed (approved):
ACK 71d1d13627ccd27319f347e2d8167c8fe8a433f4
Very nice! (one small suggestion is optional)
